### PR TITLE
added extended test for openldap image and fixtures

### PR DESCRIPTION
**Pull request by stevekuznetsov**
@deads2k @liggitt This PR will hold the fixtures and respective test. This needs https://github.com/openshift/origin/pull/4129 to merge so the `buildConfig` in this PR can access those artifacts from github.

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
[test][extended:openldap]

**Comment by deads2k** [comment]
fixes for extended merged.  [test]

File: test/extended/ldap.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// +build ldap
```
Comment (mfojtik) [inline test/extended/ldap.go]: this will be test/extended/authentication/ldap.go after https://github.com/openshift/origin/pull/4256 gets merged

Comment (mfojtik) [inline test/extended/ldap.go]: also you can nuke the // +build ldap

File: test/extended/ldap.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	. "github.com/onsi/gomega"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = Describe("OpenLDAP build and deployment", func() {
+	defer GinkgoRecover()
+	var (
+		imageStreamFixture       = filepath.Join("fixtures", "openldap", "ldapserver-imagestream.json")
```
Comment (mfojtik) [inline test/extended/ldap.go]: s/filepath.Join/exutil.FixturePath/ (https://github.com/openshift/origin/pull/4256)

File: test/extended/ldap.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		imageStreamTargetFixture = filepath.Join("fixtures", "openldap", "ldapserver-imagestream-testenv.json")
+		buildConfigFixture       = filepath.Join("fixtures", "openldap", "ldapserver-buildconfig.json")
+		deploymentConfigFixture  = filepath.Join("fixtures", "openldap", "ldapserver-deploymentconfig.json")
+		serviceConfigFixture     = filepath.Join("fixtures", "openldap", "ldapserver-service.json")
+		oc                       = exutil.NewCLI("openldap", kubeConfigPath())
+	)
+
+	Describe("Building and deploying an OpenLDAP server", func() {
```
Comment (mfojtik) [inline test/extended/ldap.go]: Describe("ldap: Building and deploying...")

File: hack/test-extended/openldap/run.sh [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+registry="$(dig @${API_HOST} "docker-registry.default.svc.cluster.local." +short A | head -n 1)"
+echo "[INFO] Registry IP - ${registry}"
+
+echo "[INFO] Starting extended tests ..."
+echo "[INFO] MASTER IP - ${MASTER_ADDR}"
+echo "[INFO] SERVER CONFIG PATH - ${SERVER_CONFIG_DIR}"
+
+KUBECONFIG="${ADMIN_KUBECONFIG}" \
+	GOPATH="${OS_ROOT}/Godeps/_workspace:${GOPATH}" \
+	go test -v -tags=ldap ./test/extended
```
Comment (mfojtik) [inline hack/test-extended/openldap/run.sh]: The launcher was refactored recently, you want to get the latest version from https://github.com/openshift/origin/pull/4256

File: test/extended/ldap.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+var _ = Describe("OpenLDAP build and deployment", func() {
+	defer GinkgoRecover()
+	var (
+		imageStreamFixture       = filepath.Join("fixtures", "openldap", "ldapserver-imagestream.json")
+		imageStreamTargetFixture = filepath.Join("fixtures", "openldap", "ldapserver-imagestream-testenv.json")
+		buildConfigFixture       = filepath.Join("fixtures", "openldap", "ldapserver-buildconfig.json")
+		deploymentConfigFixture  = filepath.Join("fixtures", "openldap", "ldapserver-deploymentconfig.json")
+		serviceConfigFixture     = filepath.Join("fixtures", "openldap", "ldapserver-service.json")
+		oc                       = exutil.NewCLI("openldap", kubeConfigPath())
```
Comment (mfojtik) [inline test/extended/ldap.go]: exutil.KubeConfigPath()

File: test/extended/ldap.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	)
+
+	Describe("Building and deploying an OpenLDAP server", func() {
+		It(fmt.Sprintf("should create a image from %q template and run it in a pod", buildConfigFixture), func() {
+			oc.SetOutputDir(testContext.OutputDir)
```
Comment (mfojtik) [inline test/extended/ldap.go]: exutil.TestContext.OutputDir

File: test/extended/ldap.go (hunk 1 above)
Comment (stevekuznetsov) [inline test/extended/ldap.go]: This test is (currently) testing the fixtures, and will in the future be expanded to also test the group sync - why `extended/authentication` ?

Comment (mfojtik) [inline test/extended/ldap.go]: https://github.com/mfojtik/origin/blob/extended-docs/docs/writing-extended-test.md#groups-vs-packages

**Comment by stevekuznetsov** [comment]
@mfojtik I cannot get this to work locally, always fails here:

```
Running Suite: OpenShift extended tests suite
=============================================
Random Seed: 1439991460
Will run 1 of 146 specs

Running in parallel across 7 nodes

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
	 -------------------------------------------------------------------
	|                                                                   |
	|  Ginkgo timed out waiting for all parallel nodes to report back!  |
	|                                                                   |
	 -------------------------------------------------------------------


Ginkgo ran 1 suite in 1.741569208s
Test Suite Failed
[INFO] Tearing down test
[INFO] Stopping k8s docker containers
<...>
[INFO] Removing k8s docker containers
<...>
[INFO] Cleanup complete
[INFO] Exiting
```

File: test/extended/ldap.go (hunk 1 above)
Comment (mfojtik) [inline test/extended/ldap.go]: @stevekuznetsov if this test is related to OpenLDAP image then maybe move it to images/ ? I can't think if a better name :)

File: hack/test-extended/ldap/run.sh [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+echo "[INFO] Registry IP - ${registry}"
+
+echo "[INFO] Starting extended tests ..."
+
+# time go test ./test/extended/ #"${OS_ROOT}/hack/listtests.go" -prefix="${OS_GO_PACKAGE}/${package}.Test" "${testdir}"  | grep --color=never -E "${1-Test}" | xargs -I {} -n 1 bash -c "exectest {} ${@:2}" # "${testexec}" -test.run="^{}$" "${@:2}"
+echo "[INFO] MASTER IP - ${MASTER_ADDR}"
+echo "[INFO] SERVER CONFIG PATH - ${SERVER_CONFIG_DIR}"
+
+# Run the tests
+KUBECONFIG="${ADMIN_KUBECONFIG}" \
```
Comment (mfojtik) [inline hack/test-extended/ldap/run.sh]: you need to pickup this change: https://github.com/openshift/origin/pull/4256/files#diff-dc02ef02d5107ceb2a902e2f46b8a16bR153

**Comment by stevekuznetsov** [comment]
@mfojtik Your change from `filepath.Join` to `exutil.FixturePath` ... there is no `FixturePath` method in my util package?

**Comment by mfojtik** [comment]
@stevekuznetsov correct, it is safe to apply that PR to your branch... should be merged today

**Comment by stevekuznetsov** [comment]
@deads2k  Takes ~200s but works!

[test][extended:ldap]

File: hack/test-extended/ldap/run.sh [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+
+cleanup() {
+	out=$?
+	set +e
+	
+	echo "[INFO] Tearing down test"
+	kill_all_processes
+	rm -rf ${ETCD_DIR-}
+	echo "[INFO] Stopping k8s docker containers"; docker ps | awk 'index($NF,"k8s_")==1 { print $1 }' | xargs -l -r docker stop
```
Comment (deads2k) [inline hack/test-extended/ldap/run.sh]: This is probably common to a lot of scripts, include the flag name right?  If so, move to `util.sh`

File: hack/test-extended/ldap/run.sh [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+		echo "[INFO] Removing k8s docker containers"; docker ps -a | awk 'index($NF,"k8s_")==1 { print $1 }' | xargs -l -r docker rm
+	fi
+
+	set -e
+	echo "[INFO] Cleanup complete"
+	echo "[INFO] Exiting"
+	exit $out
+}
+
+# Check if we have ginkgo command
```
Comment (deads2k) [inline hack/test-extended/ldap/run.sh]: Move to `util.sh` as `require_ginkgo_or_die`

File: hack/test-extended/ldap/run.sh [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+API_HOST="${API_HOST:-${DEFAULT_SERVER_IP}}"
+setup_env_vars
+mkdir -p $LOG_DIR $ARTIFACT_DIR
+
+# use the docker bridge ip address until there is a good way to get the auto-selected address from master
+# this address is considered stable
+# used as a resolve IP to test routing
+CONTAINER_ACCESSIBLE_API_HOST="${CONTAINER_ACCESSIBLE_API_HOST:-172.17.42.1}"
+
+STI_CONFIG_FILE="${LOG_DIR}/stiAppConfig.json"
```
Comment (deads2k) [inline hack/test-extended/ldap/run.sh]: check this whole file for dead vars like this.

File: hack/test-extended/ldap/run.sh (hunk 8 above)
Comment (stevekuznetsov) [inline hack/test-extended/ldap/run.sh]: Was going to leave that to https://github.com/openshift/origin/issues/4266

**Comment by stevekuznetsov** [comment]
@deads2k  Since it was using the same bucket, added test to `default` bucket. 

[test][extended:all]

**Comment by deads2k** [comment]
> @deads2k Since it was using the same bucket, added test to default bucket.

Take it out of default.  This shouldn't be a part of the basic, "do most things work".  This is a very narrow case that most people will have trouble breaking.

File: test/extended/builds/ldap_fixtures.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+		imageStreamTargetFixture = exutil.FixturePath("fixtures", "ldap", "ldapserver-imagestream-testenv.json")
+		buildConfigFixture       = exutil.FixturePath("fixtures", "ldap", "ldapserver-buildconfig.json")
+		deploymentConfigFixture  = exutil.FixturePath("fixtures", "ldap", "ldapserver-deploymentconfig.json")
+		serviceConfigFixture     = exutil.FixturePath("fixtures", "ldap", "ldapserver-service.json")
+		oc                       = exutil.NewCLI("openldap", exutil.KubeConfigPath())
+	)
+
+	Describe("Building and deploying an OpenLDAP server", func() {
+		It(fmt.Sprintf("should create a image from %q template and run it in a pod", buildConfigFixture), func() {
+			oc.SetOutputDir(exutil.TestContext.OutputDir)
```
Comment (deads2k) [inline test/extended/builds/ldap_fixtures.go]: If I were to run multiple tests in parallel, this folder would become busy, right?  Choose a unique folder.

Comment (stevekuznetsov) [inline test/extended/builds/ldap_fixtures.go]: @mfojtik Do we have a specific way we are subdividing this directory for each test?

File: test/extended/builds/ldap_fixtures.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+				},
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			By(fmt.Sprintf("calling oc create -f %q", deploymentConfigFixture))
+			err = oc.Run("create").Args("-f", deploymentConfigFixture).Execute()
+			Expect(err).NotTo(HaveOccurred())
+
+			By("starting the test environment deployment")
+			err = oc.Run("deploy").Args("openldap-server").Execute()
```
Comment (deads2k) [inline test/extended/builds/ldap_fixtures.go]: This only displays the deployment.  It doesn't actually wait for it to complete.  Wait for it to finish.

File: test/extended/builds/ldap_fixtures.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+		deploymentConfigFixture  = exutil.FixturePath("fixtures", "ldap", "ldapserver-deploymentconfig.json")
+		serviceConfigFixture     = exutil.FixturePath("fixtures", "ldap", "ldapserver-service.json")
+		oc                       = exutil.NewCLI("openldap", exutil.KubeConfigPath())
+	)
+
+	Describe("Building and deploying an OpenLDAP server", func() {
+		It(fmt.Sprintf("should create a image from %q template and run it in a pod", buildConfigFixture), func() {
+			oc.SetOutputDir(exutil.TestContext.OutputDir)
+
+			By(fmt.Sprintf("calling oc create -f %q", imageStreamFixture))
```
Comment (deads2k) [inline test/extended/builds/ldap_fixtures.go]: Call create on the entire directory at once, then wait for each condition to finish.

Comment (stevekuznetsov) [inline test/extended/builds/ldap_fixtures.go]: Why don't we want discrete steps here? That's more useful from a what-broke standpoint. no?

Comment (deads2k) [inline test/extended/builds/ldap_fixtures.go]: > Why don't we want discrete steps here? That's more useful from a what-broke standpoint. no?

Marginally, but less likely to be used if we try to describe to someone how to run the example.  The only real benefit is you have corrupted json, but even so, your search area is small.

Comment (stevekuznetsov) [inline test/extended/builds/ldap_fixtures.go]: Also, we cannot create the entire directory and then wait - we need the image stream to be ready before starting the build or else it fails.

**Comment by stevekuznetsov** [comment]
Waiting for extended test changes to die down before rebasing and re-writing.

**Comment by mfojtik** [comment]
@stevekuznetsov there are no more changes AFAIK except the shell launcher

File: test/extended/builds/ldap_fixtures.go (hunk 13 above)
Comment (deads2k) [inline test/extended/builds/ldap_fixtures.go]: > Also, we cannot create the entire directory and then wait - we need the image stream to be ready before starting the build or else it fails.

Welcome to user pain.  Does the build contorller not retry in such a case?

**Comment by deads2k** [comment]
rebase please.  Also, it doesn't seem to like all,ldap

**Comment by stevekuznetsov** [comment]
@deads2k Added test for deployment.

[test][extended]

**Comment by stevekuznetsov** [comment]
@deads2k Added test for deployment.

[test][extended]

**Comment by stevekuznetsov** [comment]
@deads2k Added test for deployment. Hopefully `extended:default` makes Jenkins happy...

[test][extended:default]

**Comment by stevekuznetsov** [comment]
Rebased to kick Jenkins, had a flake

**Comment by stevekuznetsov** [comment]
@deads2k Tests passing finally.

File: test/extended/builds/ldap_fixtures.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+		imageStreamTargetFixture = exutil.FixturePath("fixtures", "ldap", "ldapserver-imagestream-testenv.json")
+		buildConfigFixture       = exutil.FixturePath("fixtures", "ldap", "ldapserver-buildconfig.json")
+		deploymentConfigFixture  = exutil.FixturePath("fixtures", "ldap", "ldapserver-deploymentconfig.json")
+		serviceConfigFixture     = exutil.FixturePath("fixtures", "ldap", "ldapserver-service.json")
+		oc                       = exutil.NewCLI("openldap", exutil.KubeConfigPath())
+	)
+
+	g.Describe("Building and deploying an OpenLDAP server", func() {
+		g.It(fmt.Sprintf("should create a image from %s template and run it in a pod", buildConfigFixture), func() {
+			nameRegex := regexp.MustCompile(`"[A-Za-z0-9\-]+"`)
```
Comment (deads2k) [inline test/extended/builds/ldap_fixtures.go]: make an issue to switch over to `-o name` once we rebase it in.

File: test/extended/builds/ldap_fixtures.go [hunk 15]
```diff
@@ -0,0 +1,77 @@
+package builds
```
Comment (deads2k) [inline test/extended/builds/ldap_fixtures.go]: This doesn't belong in builds.  Move to authentication.

**Comment by deads2k** [comment]
This looks like a test for authentication.  It will be the base for several ldap authentication tests dealing with users and groups.  Go ahead and make the bucket.

**Comment by deads2k** [comment]
see also: https://github.com/openshift/origin/pull/4469

**Comment by stevekuznetsov** [comment]
Added to its' own bucket.

[test][extended:authentication]

**Comment by stevekuznetsov** [comment]
@deads2k Any other comments?

**Comment by stevekuznetsov** [comment]
@deads2k Added weird imports... Jenkins should run this now.

File: hack/test-extended/authentication/run.sh [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+echo "[INFO] Starting 'authentication' extended tests"
+
+dirs_image_env_setup_extended
+
+setup_env_vars
+
+trap "exit" INT TERM
+trap "cleanup_extended" EXIT
+
+info_msgs_ip_host_setup_extended
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: pretty sure this is dead.

Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: function gone.

File: hack/test-extended/authentication/run.sh [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+test_privileges
+
+echo "[INFO] Starting 'authentication' extended tests"
+
+dirs_image_env_setup_extended
+
+setup_env_vars
+
+trap "exit" INT TERM
+trap "cleanup_extended" EXIT
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: function gone.

File: hack/test-extended/authentication/run.sh [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+ginkgo_check_extended
+set +e
+
+os::build::extended
+
+test_privileges
+
+echo "[INFO] Starting 'authentication' extended tests"
+
+dirs_image_env_setup_extended
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: function gone

File: hack/test-extended/authentication/run.sh [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+source ${OS_ROOT}/hack/util.sh
+source ${OS_ROOT}/hack/common.sh
+
+set -e
+ginkgo_check_extended
+set +e
+
+os::build::extended
+
+test_privileges
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: function gone.

File: hack/test-extended/authentication/run.sh [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+echo "[INFO] Starting 'authentication' extended tests"
+
+dirs_image_env_setup_extended
+
+setup_env_vars
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: move just above configure

File: hack/test-extended/authentication/run.sh [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+trap "exit" INT TERM
+trap "cleanup_extended" EXIT
+
+info_msgs_ip_host_setup_extended
+
+configure_os_server
+
+auth_setup_extended
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: function gone

File: hack/test-extended/authentication/run.sh [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+start_os_server
+
+install_registry_extended
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: function gone.

File: hack/test-extended/authentication/run.sh [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+
+wait_for_command '[[ "$(oc get endpoints docker-registry --output-version=v1 -t "{{ if .subsets }}{{ len .subsets }}{{ else }}0{{ end }}" --config=/tmp/openshift-extended-tests/openshift.local.config/master/admin.kubeconfig || echo "0")" != "0" ]]' $((5*TIME_MIN))
```
Comment (deads2k) [inline hack/test-extended/authentication/run.sh]: superceded by built-in function

**Comment by deads2k** [comment]
other than not working, I'm probably ok with it.

**Comment by stevekuznetsov** [comment]
@deads2k Should work again... unless you introduce more changes quick :)

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
```
error: couldn't read version from server: Get https://127.0.0.1:8443/api: net/http: TLS handshake timeout
see 'oc project -h' for help.
!!! Error in hack/test-cmd.sh:239
	'oc project project-foo' exited with status 1
Call stack:
	1: hack/test-cmd.sh:239 main(...)
Exiting with status 1
```

will look for it after this meeting.

re[merge]

**Pull request merged**